src/scripting_api/util.js

```diff
@@ -20,11 +20,724 @@ class Util extends PDFObject {
+      year,
+      month,
+      day,
+      dayOfWeek,
+      hours,
+      minutes,
+      seconds,
+      milliseconds,
+      cDate,
+    };
+
+    const pattern = /(MMMM|MMM|MM|M|DD|D|YYYY|YY|JJJ|J|EEEE|EEE|E|G|w|WW|hh|h|kk|k|HH|H|KK|K|mm|m|SS|S|FFF|A|Z|zz|z|\\.)/g;
+    return cFormat.replace(pattern, function (match, p1) {
+      if (p1 in handlers) {
+        return handlers[p1](data);
+      }
+      return p1.charCodeAt(1);
+    });
+  }
+
+  printx(cFormat, cSource) {
+    // case
+    const handlers = [x => x, x => x.toUpperCase(), x => x.toLowerCase()];
+
+    // limits
+    const [LA, LZ, UA, UZ, ZER, NIN] = Array.from("azAZ09").map(c =>
+      c.charCodeAt(0)
+    );
+
+    const buf = [];
+    let i = 0;
+    const ii = cSource.length;
+    let currCase = handlers[0];
+    let escaped = false;
+
+    for (const command of cFormat) {
+      if (escaped) {
+        buf.push(command);
+        escaped = false;
+        continue;
+      }
+      if (i >= ii) {
+        break;
+      }
+      switch (command) {
+        case "?":
+          buf.push(currCase(cSource.charAt(i++)));
+          break;
+        case "X":
+          while (i < ii) {
+            const code = cSource.charCodeAt(i++);
+            if (
+              (LA <= code && code <= LZ) ||
+              (UA <= code && code <= UZ) ||
+              (ZER <= code && code <= NIN)
+            ) {
+              buf.push(currCase(String.fromCharCode(code)));
+              break;
+            }
+          }
+          break;
+        case "A":
+          while (i < ii) {
+            const code = cSource.charCodeAt(i++);
+            if ((LA <= code && code <= LZ) || (UA <= code && code <= UZ)) {
+              buf.push(currCase(String.fromCharCode(code)));
+              break;
+            }
+          }
+          break;
+        case "9":
+          while (i < ii) {
+            const code = cSource.charCodeAt(i++);
+            if (ZER <= code && code <= NIN) {
+              buf.push(String.fromCharCode(code));
+              break;
+            }
+          }
+          break;
+        case "*":
+          while (i < ii) {
+            buf.push(currCase(cSource.charAt(i++)));
+          }
+          break;
+        case "\\":
+          escaped = true;
+          break;
+        case ">":
+          currCase = handlers[1];
+          break;
+        case "<":
+          currCase = handlers[2];
+          break;
+        case "=":
+          currCase = handlers[0];
+          break;
+        default:
+          buf.push(command);
+      }
+    }
+
+    return buf.join("");
+  }
+
+  scand(cFormat, cDate) {
+    switch (cFormat) {
+      case "0":
+        return this.scand("D:yyyymmddHHMMss", cDate);
+      case "1":
+        return this.scand("yyyy.mm.dd HH:MM:ss", cDate);
+      case "2":
+        return this.scand("m/d/yy h:MM:ss tt", cDate);
+    }
+
+    if (!(cFormat in this._scandCache)) {
+      const _months = this._months;
+      const _days = this._days;
```

Comment:
Note that there's not really any precedent in the PDF.js code-base for having *local variable* names start with an underscore. (That's usually limited to indicate "private" properties/methods on e.g. objects and classes.)